### DISCUSSION TITLE
PocketBase Realtime in SwiftUI

### DISCUSSION THREAD

**Discussion by einlichtvogel** [pocketbase/pocketbase#308] [Q&A] [answered]
Hi, I wanted to implement the real-time connection in Swift UI but I cannot really get it to work.
Did someone already do it, or knows how to do it and could help me?
Maybe this is also a question for StackOverflow because it's not mainly about PocketBase.
Thanks :)

EDIT: I got it to work, and I get the message PB_CONNECT and the client id, but I do not get any messages when a thing is updated in my collection.

**Comment by ganigeorgiev** [comment, marked as answer]
Hi,

I have very little (and outdated) experience with Swift and therefore I'm not sure that I'll be able to help you, but I would try to explain the process in general.

The [Realtime  API](https://pocketbase.io/docs/api-realtime/) uses Server-sent-events (SSE) so you most likely will need a SSE library OR you can manually parse the SSE messages with a regular http client that can process a streamed response (`text/event-stream`) (this is how is done in the PocketBase Dart SDK since at the moment there isn't any well maintained SSE package - https://github.com/pocketbase/dart-sdk/blob/master/lib/src/sse/sse_client.dart).

After that, to connect to the SSE server, it is enough to send a `GET` request to `/api/realtime`.
This will start a **long-lived/persistent** connection and the first message that you'll receive will be `PB_CONNECT`:
```
id:CLIENT_ID
event:PB_CONNECT
data:{"clientId": CLIENT_ID}
```

At this stage you haven't subscribed to anything yet. To create a subscription you need to send a `POST` request to `/api/realtime` with the following `json` or `multipart/form-data` body, eg.:
```js
// authentication also happen at this stage so in addition to the body, you could send your `Authorization: User/Admin TOKEN` header
{
    "clientId": CLIENT_ID, // the client id from the PB_CONNECT message
    "subscriptions": ["collection1", "collection2/RECORD_ID", etc.]
}
```


And that's it.

Once there are any change in the subscribed collection records, you'll receive your events as another SSE message that will look like this:
```
id:CLIENT_ID
event:collection1 // this is the subscription/topic name
data:{"action": "create", "record": {...}}
```

**Reply by einlichtvogel** [reply]
Now it works, thank you!

I have already read the docs, but there for me it wasn't really clear that you have to send a push request to subscribe to it.
Do you also have to send the push request if you are using the SDK, or just if not?
That would be very nice to describe a little more precisely in the docs, if not already done and I just haven't seen it :)

**Reply by ganigeorgiev** [reply]
> Do you also have to send the push request if you are using the SDK, or just if not?

No, the SDKs handle everything seamlessly by just calling `client.realtime.subscribe()` or `client.realtime.unsubscribe()` - they will automatically start a new connection and send the subcription (or terminate an existing connection on unsubscribe if there are no other active subscriptions).

The Realtime docs will be updated in the near future (I'm planning to spend more time on this after v0.6-v0.7 release) with more in depth guides and examples for more advanced use cases.
